### PR TITLE
Converting to use System.CommandLine

### PR DESCRIPTION
**Pull request by sfoslund**

### CODE REVIEW

**Comment by sfoslund** [comment]
Left to do:
- [ ] Localize command line strings
- [ ] Audit help for commands to ensure no content changes
- [ ] Allow run command to take in tokens after "--"
- [ ] Accept case agnostic verbosity options (ie diag vs DIAG)

**Comment by sfoslund** [comment]
@jonsequitur I'm seeing any tokens after "--" being ignored, is there a way to opt out of this behavior and include those tokens in the parse result?

@wli3 tagging you here in case you have any input or feedback on the changes so far.

**Comment by AntonLapounov** [comment]
We have found that `System.CommandLine` has a noticeable performance overhead (in particular, due to its use of reflection), so we had to switch the command-line parser back in the runtime repo.

**Comment by wli3** [comment]
@AntonLapounov the new parser was a fork of existing SDK parser, so it won't be a significate regression compared with the old if not improvement. Also, CLI performance is not as sensitive as runtime. The correctness and the ability to handle complexity is higher priority.

**Comment by wli3** [comment]
@jonsequitur it uses reflection in the low abstraction API as in this case too? Or it is in invocation pipeline?

File: src/Cli/Microsoft.DotNet.Cli.Utils/DebugHelper.cs
Comment (jonsequitur) [inline src/Cli/Microsoft.DotNet.Cli.Utils/DebugHelper.cs:24]: This is supported directly by System.CommandLine using the `[debug]` directive.

File: src/Cli/dotnet/CommandBase.cs
Comment (jonsequitur) [inline src/Cli/dotnet/CommandBase.cs:25]: This looks redundant with the System.CommandLine behavior.

File: src/Cli/dotnet/CommandLineParser.cs
Comment (jonsequitur) [inline src/Cli/dotnet/CommandLineParser.cs:59]: Though the output might be slightly different, this is supported directly by System.CommandLine.

File: src/Cli/dotnet/CommandLineParser.cs
Comment (jonsequitur) [inline src/Cli/dotnet/CommandLineParser.cs:53]: For these, I would suggest using `Option<T>` rather than `Option`.

File: src/Cli/dotnet/CommandLineParser.cs
Comment (jonsequitur) [inline src/Cli/dotnet/CommandLineParser.cs:64]: I'd suggest using `Argument<T>` rather than `Argument`.  Setting the arity should be unnecessary.

File: src/Cli/dotnet/OptionForwardingExtensions.cs
Comment (jonsequitur) [inline src/Cli/dotnet/OptionForwardingExtensions.cs:43]: This can probably inherit `Option<T>`.

File: src/Cli/dotnet/Telemetry/AllowListToSendFirstAppliedOptions.cs
Comment (jonsequitur) [inline src/Cli/dotnet/Telemetry/AllowListToSendFirstAppliedOptions.cs:23]: You can also do:

```csharp
parseResult.CommandResult.Command.Name
```

File: src/Cli/dotnet/Telemetry/TelemetryFilter.cs
Comment (jonsequitur) [inline src/Cli/dotnet/Telemetry/TelemetryFilter.cs:30]: This could be an option or argument but it looks like we're expecting a command.

File: src/Cli/dotnet/Telemetry/TopLevelCommandNameAndOptionToLog.cs
Comment (jonsequitur) [inline src/Cli/dotnet/Telemetry/TopLevelCommandNameAndOptionToLog.cs:27]: This could be an option or argument but it looks like we're expecting a command.

File: src/Cli/dotnet/commands/dotnet-add/AddCommandParser.cs
Comment (jonsequitur) [inline src/Cli/dotnet/commands/dotnet-add/AddCommandParser.cs:18]: Arity should be unnecessary if you use `Argument<T>`. (You might want to use `Argument<FileInfo>` here.)

File: src/Cli/dotnet/commands/dotnet-add/dotnet-add-reference/Program.cs
Comment (jonsequitur) [inline src/Cli/dotnet/commands/dotnet-add/dotnet-add-reference/Program.cs:24]: There are overloads for `ValueForArgument` and `ValueForOption` that take `Argument<T>` and `Option<T>` respectively, so if these objects, rather than just the strings, can be referenced here, it should provide some perf benefit because the lookups are more efficient.

**Comment by sfoslund** [comment]
I ran some preliminary perf comparisons between this PR and original code and I've listed the results below. Overall, the new system.commandline version is marginally faster or comparable to the original. The exception is help, which I expected because it's the only place where we're using System.CommandLine's invoke method (since we decided to use the parser only and stick to our existing command handling for everything else).

@jonsequitur the original parser had a HelpView method for a parse result, which allowed us to display the help without having to directly invoke the command with the "-h" option. Would it be possible to have something equivalent with System.CommandLine?


Command | System.CommandLine (ms) | Original (ms)
-- | -- | --
dotnet --debug new | 3416 | 3997
dotnet --debug   tool list -g | 2290 | 2821
dotnet --debug   tool uninstall dotnet-doc -g | 2361 | 2507
dotnet --debug new   sln | 2703 | 2736
dotnet --debug new   console | 4361 | 5372
dotnet --debug sln   add test\test.csproj | 2757 | 2747
dotnet --debug add   test/test.csproj reference test1/test1.csproj | 2942 | 3266
dotnet --debug -h | 2444 | 1943

**Comment by sfoslund** [comment]
@wli3 most tests should be passing now (known failures are being tracked here: https://github.com/dotnet/sdk/issues/14461), do you mind reviewing when you have time?

File: src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs
Comment (kartheekp-ms) [inline src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs:18]: I don't see `command.AddOption(new Option(new string[] { "-h", "--help" }, CompletionOnlyDescription);` option for all the `dotnet nuget` commands. Am I missing anything?

File: src/Cli/dotnet/commands/dotnet-complete/CompleteCommand.cs
Comment (kartheekp-ms) [inline src/Cli/dotnet/commands/dotnet-complete/CompleteCommand.cs:52]:
```suggestion
var input = complete.ValueForArgument<string>(CompleteCommandParser.PathArgument) ?? string.Empty;
```

File: src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-nuget/NuGetCommandParser.cs:18]: System.CommandLine has help built in, so it's specified in the parser: https://github.com/dotnet/sdk/pull/14379/files#diff-75f24d684a51fe02f83d64c70e9cd87276f55132b61facd9fd41bb0b5d1da847R134

**Comment by kartheekp-ms** [comment]
@sfoslund - On the `NuGet Client` side, we use [`Microsoft.Extensions.CommandLineUtils`](https://www.nuget.org/packages/Microsoft.Extensions.CommandLineUtils/) package for `Command-line parsing API`. For e.g. `dotnet add package` command on our side has [this](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommand.cs) implementation. Please let us know if you are expecting any changes in the downstream due to this PR.

**Comment by sfoslund** [comment]
@kartheekp-ms you shouldn't have any effects downstream from this change. We are switching to use a different parser but the way we call you and the parameters we pass will stay the same.

**Comment by wli3** [comment]
Give me some time, it will review this thoroughly too

File: src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs
Comment (kartheekp-ms) [inline src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs:67]: The current `dotnet add package` command doesn't have `--diagnostics` option. No references to this option in NuGet side [implementation](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommand.cs) for this command. I see that this is one of the common options. Is this something new that we should support?

File: src/Cli/dotnet/commands/dotnet-list/dotnet-list-package/ListPackageReferencesCommandParser.cs
Comment (kartheekp-ms) [inline src/Cli/dotnet/commands/dotnet-list/dotnet-list-package/ListPackageReferencesCommandParser.cs:68]: The current `dotnet list package` command doesn't have `--diagnostics` option. No references to this option in NuGet side [implementation](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommand.cs) for this command. I see that this is one of the common options. Is this something new that we should support?

File: src/Cli/dotnet/Telemetry/TelemetryFilter.cs
Comment (wli3) [inline src/Cli/dotnet/Telemetry/TelemetryFilter.cs:31]: `.FirstOrDefault(c => c.Token() != null && c.Token().Type.Equals(TokenType.Command))?` happened several times. It is better to extract it to an extension method or similar

File: src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/AddPackageParser.cs:67]: Nope, we don't want to change anything that will require changes on your side in this PR, so I'll remove it. Thanks for pointing this out.

File: src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs:95]: There is no equivalent of 

```
Accept.AnyOneOf(
                            "full",
                            "mini")
``` ?

File: src/Cli/dotnet/commands/dotnet-tool/restore/ToolRestoreCommandParser.cs
Comment (wli3) [inline src/Cli/dotnet/commands/dotnet-tool/restore/ToolRestoreCommandParser.cs:14]: Same as global/local. The text between install and restore are actually different. You could try to make the text generic instead.

Please check the convert for tools command for similar issue. And you should remove unused restore/list/uninstall counterpart

**Review by wli3 — COMMENTED** [review]
not done yet :)

File: src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs:24]: This is all handled by system.commandline now, so we know the parse result will never be null and it generated the error for us if the wrong number of package references are specified.

File: src/Cli/dotnet/CommonOptions.cs
Comment (sfoslund) [reply src/Cli/dotnet/CommonOptions.cs:94]: The options in this common options file aren't included in all commands, just shared across more than one. I did include the diag option in too many commands though, I'll remove the extras- thanks for pointing this out

File: src/Cli/dotnet/commands/dotnet-list/dotnet-list-reference/ListProjectToProjectReferencesCommand.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-list/dotnet-list-reference/ListProjectToProjectReferencesCommand.cs:32]: Good catch, it doesn't look the same, I'll fix it.

File: src/Cli/dotnet/commands/dotnet-new/NewCommandParser.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-new/NewCommandParser.cs:9]: This isn't used for --help, it still goes to the template engine exe ([redirected through out help builder](https://github.com/dotnet/sdk/pull/14379/files#diff-75f24d684a51fe02f83d64c70e9cd87276f55132b61facd9fd41bb0b5d1da847R157)). But it is used for suggestions (see https://github.com/dotnet/sdk/blob/629cbd8664118463a12645b3332561ceed1b9a93/src/Tests/dotnet.Tests/CommandTests/CompleteCommandTests.cs#L84). Does the template engine exe provide suggestions that we should defer to instead?

File: src/Cli/dotnet/commands/dotnet-restore/RestoreCommandParser.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-restore/RestoreCommandParser.cs:61]: No, not new, these options used to always be included. We can't do that anymore because some commands (like build for example) already have a --runtime or --no-dependencies option. With the old parser it didn't care and would just overwrite the duplicate option, but system.commandline throws if we add an option twice.

File: src/Cli/dotnet/commands/dotnet-test/Program.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-test/Program.cs:154]: No, it wasn't being used even before this change.

File: src/Cli/dotnet/commands/dotnet-test/Program.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-test/Program.cs:167]: Yes, we get this help behavior for free with system.commandline

File: src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs:95]: There is, FromAmong, which I'm using with the options above and below this one. Should this option only accept full or mini as well? The original code doesn't specify that.

File: src/Cli/dotnet/commands/RestoringCommand.cs
Comment (wli3) [inline src/Cli/dotnet/commands/RestoringCommand.cs:56]: the existing logic does not have "-target:Restore". And usually ` msbuild -target:Restore;Build` is a bug since restore will include more props. `msbuild /restore -target:build` is the right way

File: src/Cli/dotnet/Program.cs
Comment (wli3) [inline src/Cli/dotnet/Program.cs:85]: we removed this type of error report?

File: src/Cli/dotnet/Program.cs
Comment (wli3) [inline src/Cli/dotnet/Program.cs:155]: yay! we don't have 2 layers of parsers anymore!

File: src/Cli/dotnet/Program.cs
Comment (wli3) [inline src/Cli/dotnet/Program.cs:161]: equals works here? Does system commandline have special handing for command being equal?

File: src/Cli/dotnet/Program.cs
Comment (wli3) [inline src/Cli/dotnet/Program.cs:223]: nit: some lines are too long in this change.

**Review by wli3 — COMMENTED** [review]
Finished reviewing all production code. Now looking into tests...

File: src/Cli/dotnet/commands/dotnet-new/NewCommandParser.cs
Comment (wli3) [reply src/Cli/dotnet/commands/dotnet-new/NewCommandParser.cs:9]: no, so NewCommandParser is only used for suggestion not help even after system.commandline change?

That should be fine, one day template engine would use system commandline too, so these work are not wasted

Comment (sfoslund) [reply src/Cli/dotnet/commands/dotnet-new/NewCommandParser.cs:9]: > no, so NewCommandParser is only used for suggestion not help even after system.commandline change?

Right, only used for suggestions.

File: src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs
Comment (wli3) [inline src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs:40]: It has errors now? What are they?

**Review by wli3 — COMMENTED** [review]
now I am done for an round of review

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs
Comment (sfoslund) [reply src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:68]: This is another instance of https://github.com/dotnet/command-line-api/issues/1077, the environment variables are specified after an option flag so they are being treated as option arguments, which is incorrect.

File: src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs
Comment (sfoslund) [reply src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs:40]: System.CommandLine reports any errors if there are unmatched tokens. So `dotnet tool run dotnetsay -- hi` is fine but `dotnet tool run dotnetsay hi` causes an error.

File: src/Cli/dotnet/commands/RestoringCommand.cs
Comment (sfoslund) [reply src/Cli/dotnet/commands/RestoringCommand.cs:56]: This logic existed before, just in a different location. Before, we called RestoreCommand.FromArgs (see line 60) to get the restore command whereas now we're using the RestoreCommand constructor directly (because we don't want to have to re-parse everything). The FromArgs method adds "-target:Restore" to the arguments whereas the constructor doesn't.

File: src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs
Comment (wli3) [reply src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:68]: so this feature does not work? Could you work with dotnet test team to see if there is any workaround?

Comment (sfoslund) [reply src/Tests/dotnet-test.Tests/GivenDotnetTestContainsEnvironmentVariables.cs:68]: It will work once dotnet/command-line-api#1077 is resolved :)

File: src/Cli/dotnet/Program.cs
Comment (wli3) [reply src/Cli/dotnet/Program.cs]: so i assume system.commandline has special handing for command being equal

File: src/Tests/dotnet-run.Tests/GivenDotnetRunRunsCsProj.cs
Comment (wli3) [inline src/Tests/dotnet-run.Tests/GivenDotnetRunRunsCsProj.cs:569]: is -- required now? It used to be optional

File: src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs
Comment (wli3) [reply src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs:40]: similar to the other comment. it will still run right? We disabled the error when it is unmatched right?

**Comment by wli3** [comment]
All issues I find should be resolved now. Other than the 2 new questions. If the answers are we are good. Let's merge it before it get more merge conflicts

**Review by wli3 — APPROVED** [review]
All issues I find should be resolved now. Other than the 2 new questions. If the answers are we are good. Let's merge it before it get more merge conflicts

File: src/Cli/dotnet/Program.cs
Comment (sfoslund) [reply src/Cli/dotnet/Program.cs]: Not that I'm aware of, I changes this to check that the names are equal in one of my later commits.

File: src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs
Comment (sfoslund) [reply src/Tests/dotnet.Tests/ParserTests/ToolRunParserTests.cs:40]: Yes, we have `TreatUnmatchedTokensAsErrors` set to false for both the run commands

File: src/Tests/dotnet-run.Tests/GivenDotnetRunRunsCsProj.cs
Comment (sfoslund) [reply src/Tests/dotnet-run.Tests/GivenDotnetRunRunsCsProj.cs]: Just switched it back to being optional: https://github.com/dotnet/sdk/pull/14379/commits/5927d4ccad0aa47cfccb95ba5c8a6b6b0e6decbe#diff-c59ee7ae48403cce44671ddc3549eb921e7c23bda26840c883b29d9f7e7942e5R36

**Review by wli3 — APPROVED** [review]

**Pull request merged**
